kernel/src/log_segment.rs

```diff
@@ -0,0 +1,489 @@
+        self
+    }
+
+    pub(crate) fn with_start_version(mut self, version: Version) -> Self {
+        let _ = self.start_version.insert(version);
+        self
+    }
+    pub(crate) fn with_end_version(mut self, version: Version) -> Self {
+        let _ = self.end_version.insert(version);
+        self
+    }
+    pub(crate) fn with_no_checkpoint_files(mut self) -> Self {
+        self.no_checkpoint_files = true;
+        self
+    }
+    pub(crate) fn with_in_order_commit_files(mut self) -> Self {
+        self.in_order_commit_files = true;
+        self
+    }
+    pub(crate) fn build(self) -> DeltaResult<LogSegment> {
+        let Self {
+            fs_client,
+            log_root,
+            checkpoint,
+            start_version,
+            end_version,
+            no_checkpoint_files,
+            in_order_commit_files,
+        } = self;
+        let log_url = log_root.join("_delta_log/").unwrap();
+        let (mut commit_files, mut checkpoint_files) = match (checkpoint, end_version) {
+            (Some(cp), None) => {
+                Self::list_log_files_with_checkpoint(&cp, fs_client.as_ref(), &log_url)?
+            }
+            (Some(cp), Some(version)) if cp.version >= version => {
+                Self::list_log_files_with_checkpoint(&cp, fs_client.as_ref(), &log_url)?
+            }
+            _ => Self::list_log_files(fs_client.as_ref(), &log_url)?,
+        };
+
+        // Remove checkpoint files
+        if no_checkpoint_files {
+            checkpoint_files.clear();
+        }
+
+        // Commit file versions must satisfy the following:
+        // - Must be greater than the start version
+        // - Must be greater than the most recent checkpoint version if it exists
+        // - Must be less than or equal to the specified end version.
+        if let Some(end_version) = end_version {
+            commit_files.retain(|log_path| log_path.version <= end_version);
+        }
+        if let Some(checkpoint_file) = checkpoint_files.first() {
+            commit_files.retain(|log_path| checkpoint_file.version < log_path.version);
+        }
+        if let Some(start_version) = start_version {
+            commit_files.retain(|log_path| start_version <= log_path.version);
+        }
+
+        // get the effective version from chosen files
+        let version_eff = commit_files
+            .last()
+            .or(checkpoint_files.first())
+            .ok_or(Error::MissingVersion)? // TODO: A more descriptive error
+            .version;
+        if let Some(end_version) = end_version {
+            require!(
+                version_eff == end_version,
+                Error::MissingVersion // TODO more descriptive error
+            );
+        }
+
+        // We assume listing returned ordered. If `in_order_commit_files` is false, we want reverse order.
+        if !in_order_commit_files {
+            // We assume listing returned ordered, we want reverse order
+            commit_files.reverse();
```

Comment:
Assuming listing returns reversed order, so you want just non-reversed order? Sorry nothing wrong (big tables might not like the perf here) but this just reads in a bit of a confusing way to me.

Reply:
Resolved in #495